[Security][Workflows][9.5 & Serverless]  Document `waitForApproval` step and HITL external delivery (#7509)
elastic/docs-content: explore-analyze/workflows/authoring-techniques/human-in-the-loop.md

## The mechanism: `waitForInput` and `waitForApproval`

HITL is built on wait steps that pause execution until a human responds:

| Step | Use when | Response shape |
|---|---|---|
| [`waitForInput`](/explore-analyze/workflows/steps/wait-for-input.md) | You need a custom form (notes, severity, multi-field decisions) | Your JSON Schema payload (see [output shape](#hitl-output-shape)) |
| [`waitForApproval`](/explore-analyze/workflows/steps/wait-for-approval.md) {applies_to}`stack: preview 9.5+` {applies_to}`serverless: preview` | The decision is approve or reject | `approved: true` or `false` under `output.response` |

When the workflow reaches either step, execution pauses in the `WAITING_FOR_INPUT` state. The responder sees the message and form (or approve/reject controls). When they respond, the workflow resumes with their input available as `steps.<step_name>.output`.

## What happens while the workflow is paused

While a HITL step is waiting, the execution status is `WAITING_FOR_INPUT` and the run appears in the execution history with a resume action.

> {applies_to}`stack: preview 9.5+` {applies_to}`serverless: preview` When Inbox is enabled, the waiting action also appears there.

Reviewer: ```suggestion
{applies_to}`stack: preview 9.5+` {applies_to}`serverless: preview` When [Inbox](#workflows-hitl-inbox) is enabled, the waiting action also appears there.
```

Author: Addressed in [`3586491` (this PR)](https://github.com/elastic/docs-content/pull/7509/changes/35864915eb48903dc708aa5ce9fde7270050b7c5).

Timeout behavior depends on the {{stack}} version that you're using:

::::{applies-switch}

:::{applies-item} { stack: preview 9.5+, serverless: preview }

> Wait steps time out by default: `waitForInput` after `72h` and `waitForApproval` after `24h`. These defaults apply whether or not you configure an external channel. Override them with a top-level `timeout` on the step. If no one responds before the timeout, the step fails.

Reviewer: > If no one responds before the timeout, the step fails.

This sounds the same as the statement on line 58, just worded differently. I think the sentence on line 58 should be inside the `stack: ga 9.4` applies-item?

Author: Good catch - I restructured this into an `applies-switch` so the `stack: ga 9.4` behavior (no default timeout; waits until someone responds, or until a workflow-level `settings.timeout` cancels the run) is now separate from the 9.5+ default-timeout behavior, removing the duplication. Addressed in [`3586491` (this PR)](https://github.com/elastic/docs-content/pull/7509/changes/35864915eb48903dc708aa5ce9fde7270050b7c5).

  - name: record_decision
    type: cases.addComment
    with:
      case_id: "{{ steps.open_case.output.case.id }}"
      comment: |
        **Decision:** {% if steps.review.output.response.approved %}isolated{% else %}no action{% endif %}
        **Notes:** {{ steps.review.output.response.notes }}
        **Responded by:** {{ steps.review.output.respondedBy }}
```

Execution pauses at `review`. Until someone responds, the execution state is `WAITING_FOR_INPUT`. When they respond, execution resumes at `isolate`, which is gated by an `if` guard on the approval decision.

### Simplify a yes/no decision with `waitForApproval`

```{applies_to}
stack: preview 9.5+
serverless: preview
```

The `review` step above is a simple yes/no decision, so you can replace it with [`waitForApproval`](/explore-analyze/workflows/steps/wait-for-approval.md), which renders approve/reject controls without a schema:

```yaml
  - name: review
    type: waitForApproval
    with:
      message: |
        ## Alert on `{{ event.alerts[0].host.name }}`

        **AI classification:** {{ steps.classify.output.category }}

        Isolate this host?
```

> Downstream steps read the decision from `{{ steps.review.output.response.approved }}`, exactly as they would with `waitForInput`.

Reviewer 2: **Medium — wording nit.** “Exactly as they would with `waitForInput`” is only true when the prior `waitForInput` schema used a boolean field named `approved`. Otherwise the downstream path differs (`response.<customField>` vs `response.approved`).

Suggest: “Downstream steps read `{{ steps.review.output.response.approved }}` — the same path as a `waitForInput` schema that defines an `approved` boolean.”

Author: Addressed in [`3586491` (this PR)](https://github.com/elastic/docs-content/pull/7509/changes/35864915eb48903dc708aa5ce9fde7270050b7c5).

## Resume a paused workflow

Resume a paused workflow using the following methods.

### From the Kibana UI

Open the execution view. The paused step renders a form generated from the `schema` (or approve/reject controls for `waitForApproval`). Fill it in, submit, and the workflow resumes.

> ### From the Inbox app [workflows-hitl-inbox]

Reviewer: When enabled, would the Inbox app also appear in the Kibana UI? If so, we may want to rename the previous section to "From the execution view".
Should we explain how/where users would access the Inbox app?

Author: Good points. I renamed the previous section to **From the execution view**, and clarified that Inbox is a separate Kibana app that you open from the navigation menu. Addressed in [`3586491` (this PR)](https://github.com/elastic/docs-content/pull/7509/changes/35864915eb48903dc708aa5ce9fde7270050b7c5).

```{applies_to}
stack: preview 9.5+
serverless: preview
```

The **Inbox** app lists HITL actions that need a response. Enable it with `xpack.inbox.enabled: true` in `kibana.yml` (disabled by default).

Inbox splits into:

- **Awaiting response** — Pending `waitForInput` / `waitForApproval` steps. Open the **Respond** flyout to fill in the form (or approve/reject). If the preceding step included a reasoning summary, Inbox shows it to help the responder decide.
- **History** — A record of actions that were responded to, timed out, or whose workflow was deleted. Each entry includes who responded, their response, the channel, and when it happened.

> If more than one person responds, only the first response is accepted.

Reviewer: I'm a bit confused as to how more than one person can respond. Does the workflow not resume as soon as one response is provided?

Also, does this principle only apply when responding from the Inbox app, or would it also apply when responding using the other listed methods?

Author: Yeah, the workflow resumes as soon as the first response comes in. I moved this into the **Resume a paused workflow** intro and reworded it to clarify that the step resumes on the first response, the resume token is single-use, and this applies across all resume methods (not just Inbox). Addressed in [`3586491` (this PR)](https://github.com/elastic/docs-content/pull/7509/changes/35864915eb48903dc708aa5ce9fde7270050b7c5).

### From the API

Send a `POST` request to the resume endpoint with the responder's input:

```http

> POST /api/workflowExecutions/{executionId}/resume

Reviewer 2: **Blocker (correctness vs Kibana impl):** This resume API example is wrong on both path and body.

Current docs:
```http
POST /api/workflowExecutions/{executionId}/resume

{
  "approved": true,
  "notes": "..."
}
```

Actual contract (`workflows_management` `POST /api/workflows/executions/{executionId}/resume`):
```http
POST /api/workflows/executions/{executionId}/resume
Content-Type: application/json

{
  "input": {
    "approved": true,
    "notes": "Confirmed malicious. Proceeding with isolation."
  }
}
```

For `waitForApproval`, the resume body is `{ "input": { "approved": true|false } }` — not top-level fields.

This was already incorrect before this PR; removing the author note that questioned the shape makes it more important to fix here.

Reviewer 2: @rosomri +

Author: Fixed in [`3586491` (this PR)](https://github.com/elastic/docs-content/pull/7509/changes/35864915eb48903dc708aa5ce9fde7270050b7c5).

- Path corrected to `/api/workflows/executions/{executionId}/resume` 
- Body now wraps fields in the `input` object 
- Intro sentence was updated to say "wrapped in an `input` object," and a `Content-Type` header was added, matching the exact contract `h88` cited.

@h88 just so I know whether this change need to be tagged with a specific version was the route to the `resume` endpoint updated in a previous version, or has [this example](https://github.com/elastic/docs-content/pull/7509/changes#diff-083694e5a0727b634e2d8237b058d70c892bb07cb5d49d8b9176e0519bdd06acL125) always been wrong?

Reviewer 2: it has always been wrong in production...

Reviewer 2: Explanation: The route was updated while the feature PR was still in progress, after the docs had already been generated. As a result, these docs reflected the old route, but that version was never released to production. So there wasn't a previous released version with the incorrect route... and this change simply brings the generated docs back in sync with the final (9.5) implementation

}
```

The input body becomes part of the step output. See [Output shape](#hitl-output-shape) for how to reference fields by Stack version.

### From an external channel (Slack) [workflows-hitl-external-channels]

```{applies_to}
stack: preview 9.5+
serverless: preview
```

> Configure `with.channels` on `waitForInput` or `waitForApproval` to notify responders in Slack. Slack is the only built-in channel in 9.5.

Reviewer: ```suggestion
Configure `with.channels` on `waitForInput` or `waitForApproval` to notify responders in Slack. Slack is the only supported built-in channel.
```
Suggestion to avoid hard-coding versions.

Author: Addressed in [`3586491` (this PR)](https://github.com/elastic/docs-content/pull/7509/changes/35864915eb48903dc708aa5ce9fde7270050b7c5).

| Channel key | What you need |
|---|---|
| `slack` | A Slack **webhook** connector. Posts to the webhook's configured channel. |

> | `slack_api` | A Slack **API** connector plus one or more Slack channel IDs (`channels: ["C…"]`). |

Reviewer: I wonder if it might be better to link to https://docs-v3-preview.elastic.dev/elastic/docs-content/pull/7509/explore-analyze/workflows/steps/wait-for-input#external-channels and https://docs-v3-preview.elastic.dev/elastic/docs-content/pull/7509/explore-analyze/workflows/steps/wait-for-approval#external-channels instead of having another config table here, for a couple of reasons:
* Including the config table here moves this section into the "how to configure" territory, when the main focus is about "how to resume"
* If more channels are added in future releases, we'd end up maintaining the config tables in both places.

Author: Agreed — I removed the duplicate config table here and instead link to the `waitForInput` and `waitForApproval` external-channels sections, so this section stays focused on "how to resume" and we only maintain the channel config in one place. Addressed in [`3586491` (this PR)](https://github.com/elastic/docs-content/pull/7509/changes/35864915eb48903dc708aa5ce9fde7270050b7c5).

How responders act:

- **`waitForApproval`** — Approve and reject links (and Slack API buttons) that resume the step. Responders can act without signing in to {{kib}}.

> - **`waitForInput`** — A link to a hosted form where responders submit the schema fields. The Slack notification can also include a one-click query link.

Reviewer: I find these descriptions a bit too compressed. Adding a suggestion to turn them into full sentences.

```suggestion
- **`waitForApproval`** — Responders receive approve and reject links (and Slack API buttons) that allow them to resume the step. They can act without signing in to {{kib}}.
- **`waitForInput`** — Responders receive a link to a hosted form where they submit the schema fields. The Slack notification can also include a one-click query link.
```

Author: Addressed in [`3586491` (this PR)](https://github.com/elastic/docs-content/pull/7509/changes/35864915eb48903dc708aa5ce9fde7270050b7c5).

External links work without a {{kib}} session. Each link includes a short-lived, single-use token that {{kib}} invalidates after use, timeout, or workflow cancellation.

:::{warning}
External channels send public, short-lived resume links. Don't use them for destructive, production-impacting, or hard-to-reverse workflows.
:::

> To resume waiting workflows from {{kib}} only, set external resume to `false` in `kibana.yml`:

Reviewer: Does "from {{kib}} only" mean "only from the [Kibana UI](https://docs-v3-preview.elastic.dev/elastic/docs-content/pull/7509/explore-analyze/workflows/authoring-techniques/human-in-the-loop#from-the-kibana-ui)" or does it include the Inbox app and API methods? If it includes all three methods, and only excludes the external channel method, it would be clearer to rephrase to something like "To disable the external channels option...".

Author: Clarified — I rephrased this to spell out that turning off external resume means workflows can only be resumed from {{kib}} (the execution view, the Inbox app, or the resume API); only the external channel resume links are disabled. Addressed in [`3586491` (this PR)](https://github.com/elastic/docs-content/pull/7509/changes/35864915eb48903dc708aa5ce9fde7270050b7c5).

```yaml
workflowsManagement.hitlExternalResume.enabled: false
workflowsExecutionEngine.hitlExternalResume.enabled: false
```

Both settings default to `true`. When set to `false`, Slack and other external resume links stop working.

## Output shape [hitl-output-shape]

How you read the resume payload depends on the {{stack}} version:

- {applies_to}`stack: preview 9.5+` {applies_to}`serverless: preview` Fields are under `output.response`, and the responder is `output.respondedBy`. For example: `{{ steps.review.output.response.approved }}`, `{{ steps.review.output.response.notes }}`, and `{{ steps.review.output.respondedBy }}`.
- {applies_to}`stack: ga 9.4` Fields are on the step output directly. For example: `{{ steps.review.output.approved }}` and `{{ steps.review.output.notes }}`.

## Design a good HITL form

A HITL message is read by a human mid-incident. Design for speed:

- **Lead with the decision.** The first line should say what the responder needs to decide.
- **Include the evidence.** Relevant context (alert details, enrichment results, AI rationale) belongs in the message so the responder doesn't have to dig.

> - **Keep the schema small.** Three fields is a lot. One boolean plus an optional notes field is often enough. Prefer [`waitForApproval`](/explore-analyze/workflows/steps/wait-for-approval.md) {applies_to}`stack: preview 9.5+` {applies_to}`serverless: preview` when you only need yes/no.

Reviewer: ```suggestion
- **Keep the schema small.** Three fields is a lot. One boolean plus an optional notes field is often enough. Prefer [`waitForApproval`](/explore-analyze/workflows/steps/wait-for-approval.md) when you only need yes/no.
```
The mid-sentence applies_to tag makes it difficult to read. Since we're linking to the step page that sets the applicability context, I think we can remove the tag here.

Author: Addressed in [`3586491` (this PR)](https://github.com/elastic/docs-content/pull/7509/changes/35864915eb48903dc708aa5ce9fde7270050b7c5).
